ports/rp2: simplify allocation of GC heap.

cpottle9 opened the pull request:
Borrowing an idea from the mimxrt port to simplify allocation of GC heap

In the loader input file memmap_mp.ld calculate
__GcHeapStart and __GcHeapEnd as the unused RAM.
In main.c use these addresses as arguments to gc_init().

The benefits of this change are:

1) When libraries are added or removed in the future changing BSS usage main.c does not need to be changed.

2) Currently these changes make the GC area about 30 KBytes larger. Without that change this RAM would never get used.

3) If someone wants to disable one or more SRAM blocks on the RP2040 to reduce power consumption it will be easy. Just change the MEMORY section in memmap_mp.ld.
For instance to not use SRAM2 and SRAM3 change it to:

>MEMORY
>{
>    FLASH(rx) : ORIGIN = 0x10000000, LENGTH = 2048k
>    RAM(rwx) : ORIGIN =  0x21000000, LENGTH = 128k
>    SCRATCH_X(rwx) : ORIGIN = 0x20040000, LENGTH = 4k
>    SCRATCH_Y(rwx) : ORIGIN = 0x20041000, LENGTH = 4k
>}

To turn off clocks for SRAM2 and SRAM3 from micropython set the appropriate bits in WAKE_EN0 and SLEEP_EN0.

Tested by running the firmware.uf2 file and displaying micropython.mem_info().
Confirmed GC total size approximately matched the size calculated by the loader.

cpottle9 committed:
ports/rp2: Simplify allocation of GC heap.

Borrowing an idea from the mimxrt port.
In the loader input file memmap_mp.ld calculate
__GcHeapStart and __GcHeapEnd as the unused RAM.
In main.c use these addresses as arguments to gc_init().

The benefits of this change are:

1) When libraries are added or removed in the future
changing BSS usage main.c does not need to be changed.

2) Currently these changes make the GC area about 30 KBytes
larger. Without that change this RAM would never get used.

3) If someone wants to disable one or more SRAM blocks
on the RP2040 to reduce power consumption it will be easy.
Just change the MEMORY section in memmap_mp.ld.
For instance to not use SRAM2 and SRAM3 change it to:

>MEMORY
>{
>    FLASH(rx) : ORIGIN = 0x10000000, LENGTH = 2048k
>    RAM(rwx) : ORIGIN =  0x21000000, LENGTH = 128k
>    SCRATCH_X(rwx) : ORIGIN = 0x20040000, LENGTH = 4k
>    SCRATCH_Y(rwx) : ORIGIN = 0x20041000, LENGTH = 4k
>}

To turn off clocks for SRAM2 and SRAM3 from micropython set
the appropriate bits in WAKE_EN0 and SLEEP_EN0.

Tested by running the firmware.uf2 file and displaying
micropython.mem_info().
Confirmed GC total size approximately matched the size calculated
by the loader.

cpottle9 committed:
ports/rp2: Simplify allocation of GC heap.

Borrowing an idea from the mimxrt port.
In the loader input file memmap_mp.ld calculate
__GcHeapStart and __GcHeapEnd as the unused RAM.
In main.c use these addresses as arguments to gc_init().

The benefits of this change are:

1) When libraries are added or removed in the future
changing BSS usage main.c does not need to be changed.

2) Currently these changes make the GC area about 30 KBytes
larger. Without that change this RAM would never get used.

3) If someone wants to disable one or more SRAM blocks
on the RP2040 to reduce power consumption it will be easy.
Just change the MEMORY section in memmap_mp.ld.
For instance to not use SRAM2 and SRAM3 change it to:

>MEMORY
>{
>    FLASH(rx) : ORIGIN = 0x10000000, LENGTH = 2048k
>    RAM(rwx) : ORIGIN =  0x21000000, LENGTH = 128k
>    SCRATCH_X(rwx) : ORIGIN = 0x20040000, LENGTH = 4k
>    SCRATCH_Y(rwx) : ORIGIN = 0x20041000, LENGTH = 4k
>}

To turn off clocks for SRAM2 and SRAM3 from micropython set
the appropriate bits in WAKE_EN0 and SLEEP_EN0.

Tested by running the firmware.uf2 file and displaying
micropython.mem_info().
Confirmed GC total size approximately matched the size calculated
by the loader.

cpottle9 committed:
ports/rp2 simplify allocation of GC heap

Borrowing an idea from the mimports/rp2 simplify allocation of GC heap

Borrowing an idea from the mimxrt port.
In the loader input file memmap_mp.ld calculate
__GcHeapStart and __GcHeapEnd as the unused RAM.
In main.c use these addresses as arguments to gc_init().

The benefits of this change are:

1) When libraries are added or removed in the future
changing BSS usage main.c does not need to be changed.

2) Currently these changes make the GC area about 30 KBytes
larger. Without that change this RAM would never get used.

3) If someone wants to disable one or more SRAM blocks
on the RP2040 to reduce power consumption it will be easy.
Just change the MEMORY section in memmap_mp.ld.
For instance to not use SRAM2 and SRAM3 change it to:

>MEMORY
>{
>    FLASH(rx) : ORIGIN = 0x10000000, LENGTH = 2048k
>    RAM(rwx) : ORIGIN =  0x21000000, LENGTH = 128k
>    SCRATCH_X(rwx) : ORIGIN = 0x20040000, LENGTH = 4k
>    SCRATCH_Y(rwx) : ORIGIN = 0x20041000, LENGTH = 4k
>}

To turn off clocks for SRAM2 and SRAM3 from micropython set
the appropriate bits in WAKE_EN0 and SLEEP_EN0.

Tested by running the firmware.uf2 file and displaying
micropython.mem_info().
Confirmed GC total size approximately matched the size calculated
by the loader.

dpgeorge commented:
Thanks for the contribution.  This is how stm32 has worked from the very beginning, but some ports since then have had fixed-size heaps.

One drawback of a flexible heap size (use as much RAM as possible) is that some code/apps may rely on there being a certain amount, and then in a next release of the MicroPython firmware there is a little less, and the app stops running.  But that's a small drawback for a big benefit of increased heap size for everyone.  And stm32 has had no problems with this flexible heap.

So I think this PR is a good improvement.

cpottle9 committed:
Merge branch 'simplify-gcheap' of https://github.com/cpottle9/micropython-gcheap into simplify-gcheap

cpottle9 committed:
ports/rp2: Simplify allocation of GC heap.

Borrowing an idea from the mimxrt port.
In the loader input file memmap_mp.ld calculate
__GcHeapStart and __GcHeapEnd as the unused RAM.
In main.c use these addresses as arguments to gc_init().

The benefits of this change are:

1) When libraries are added or removed in the future
changing BSS usage main.c does not need to be changed.

2) Currently these changes make the GC area about 30 KBytes
larger. Without that change this RAM would never get used.

3) If someone wants to disable one or more SRAM blocks
on the RP2040 to reduce power consumption it will be easy.
Just change the MEMORY section in memmap_mp.ld.
For instance to not use SRAM2 and SRAM3 change it to:

>MEMORY
>{
>    FLASH(rx) : ORIGIN = 0x10000000, LENGTH = 2048k
>    RAM(rwx) : ORIGIN =  0x21000000, LENGTH = 128k
>    SCRATCH_X(rwx) : ORIGIN = 0x20040000, LENGTH = 4k
>    SCRATCH_Y(rwx) : ORIGIN = 0x20041000, LENGTH = 4k
>}

To turn off clocks for SRAM2 and SRAM3 from micropython set
the appropriate bits in WAKE_EN0 and SLEEP_EN0.

Tested by running the firmware.uf2 file and displaying
micropython.mem_info().
Confirmed GC total size approximately matched the size calculated
by the loader.

cpottle9 commented:
I giveup. 
I'm new to github.
I've been trying to resolve the commit formatting with no luck.
I'm going to be like the xkcd cartoon and startover.
You will see a new pull request in a day.

dpgeorge commented:
@cpottle9 don't worry too much about the commit formatting CI failure.  It is easier for maintainers if all the CI passes, but that one in particular can be easily fixed during rebase/merge (and usually is anyway).

cpottle9 commented:
Damien, as a first time submitter I don't know what the policies are and I don't know how long PR's that are acceptable will wait to be merged.
I thought the commit formatting was blocking my PR being merged.

In the end, it was my fault because I did not set up uncrustify and the pre-commit checks on my machine.
I did now, so I should not run into this problem in the future.

In the end, starting over is not a problem.
